lib/puppet/rest/client.rb

```diff
@@ -0,0 +1,97 @@
+require 'httpclient'
+
+require 'puppet'
+require 'puppet/rest/response'
+require 'puppet/rest/server_resolver'
+
+module Puppet::Rest
+  class Client
+    def self.default_client
+      HTTPClient.new(
+        agent_name: nil,
+        default_header: {
+          'User-Agent' => Puppet.settings[:http_user_agent],
+          'X-PUPPET-VERSION' => Puppet::PUPPETVERSION
+        })
+    end
+
+    attr_reader :base_url
+
+    # Create a new HTTP client for querying the given API.
+    # @param [Puppet::Rest::Route] route data about the API being queried,
+    #                              including the API name and server details.
+    # @param [OpenSSL::X509::Store] ssl_store the SSL configuration for this client
+    # @param [Integer] receive_timeout how long in seconds this client will wait
+    #                  for a response after making a request
+    # @param [HTTPClient] client the third-party HTTP client wrapped by this
+    #                     class. This param is only used for testing.
+    # @param [Puppet::Rest::ServerResolver] server_resolver the object responsible
+    #                                       for finding the best available server
+    #                                       given the data in `route`. This param
+    #                                       is only used for testing.
+    def initialize(route,
+                   ssl_store: OpenSSL::X509::Store.new,
+                   receive_timeout: 3600,
+                   client: Puppet::Rest::Client.default_client,
+                   server_resolver: Puppet::Rest::ServerResolver.new)
+      @client = client
+
+      @server_resolver = server_resolver
+      configure_client(ssl_store, route, receive_timeout)
+    end
+
+    # Configures the underlying HTTPClient
+    # @param [OpentSSL::X509::Store] ssl_store the SSL configuration for this client
+    # @param [Puppet::Rest::Route] route data about the API being queried
+    # @param [Integer] timeout how long to wait for a response from the server once
+    #                  a request has been made
+    def configure_client(ssl_store, route, timeout)
+      @client.tcp_keepalive = true
+      @client.connect_timeout = 10
+      @client.receive_timeout = timeout
+      # `request_filter` is a list of objects implementing the `filter_request` and
+      # `filter_response` methods, which get called during request processing.
+      @client.request_filter << self
+
+      @client.cert_store = ssl_store
+
+      server, port = server_and_port(route)
+      @base_url = "https://#{server}:#{port}#{route.api}"
```

Comment:
`api` would be like `/puppet/v3`, right? And during a normal client run we'll only really be hitting that puppet endpoint. Though for the `/puppet-ca/v1` endpoint we'd want another client with specialized connection options during pki bootstrap, right? Is there any point when we'd want to keep a client around with different api prefixes?

Reply:
I don't think there's a reason to keep both around, though I do think it's cleanest to use two different ones for the different APIs, and maybe even two different ones for downloading certs vs. CRLs, because of the different security configuration.

Reply:
I dug through the base_url method for http_client, and just found this little [snippet](https://github.com/nahi/httpclient/blob/master/lib/httpclient.rb#L407-L409) that would mean we would need to make sure `route.api` returns an ending `/`. 

```
  # The last '/' is important for base_url:
  #   1. http://foo/bar/baz/ + path -> http://foo/bar/baz/path
  #   2. http://foo/bar/baz + path -> http://foo/bar/path
```

Reply:
Ah okay. That's different than how we generally present it in our docs (for example https://puppet.com/docs/puppet/5.3/http_api/http_certificate.html, which has the leading slashes on the endpoints), but I think that's alright, since this part is all internal. I'll update.